main/SQLHelper.cpp

```diff
@@ -5130,6 +5145,158 @@ void CSQLHelper::UpdateMeter()
 		pTypeGeneral, sTypeCounterIncremental,
 		pTypeGeneral, sTypeKwh
 		);
+	std::string result(buff);
+	return result;
+}
+
+bool CSQLHelper::UpdateCalendarMeter(const int HardwareID, const char* DeviceID, const unsigned char unit, const unsigned char devType, const unsigned char subType, float value,  const char* date, float usage, float counter, float min, float max, const bool dayCalendar, const char* clearAfterDate, const char* clearBeforeDate)
+{
+	std::vector<std::vector<std::string> > result;
+	std::string filter = getMeterFilter();
+	char szTmp[200];
+	
+	if((date == NULL) || (date[0] == '\0')) {
+		return false;
+	}
+	
+	result = safe_query("SELECT ID FROM DeviceStatus WHERE (HardwareID=%d AND DeviceID='%q' AND Unit=%d AND Type=%d AND SubType=%d AND (%s))", 
+		HardwareID, DeviceID, unit, devType, subType, filter.c_str());
+		
+	if (result.empty()) {
+		return false;
+	}
+	
+	uint64_t DeviceRowID;
+
+	std::vector<std::string> sd = result[0];
+	std::stringstream s_strid;
+	s_strid << sd[0];
+	s_strid >> DeviceRowID;
+	
+	if (dayCalendar)
+	{
+		if((clearBeforeDate != NULL) && (clearBeforeDate[0] != '\0') && (clearAfterDate != NULL) && (clearAfterDate[0] != '\0')) {
+			safe_query("DELETE FROM Meter WHERE (DeviceRowID=='%" PRIu64 "') AND (Date<='%q') AND (Date<='%q')", DeviceRowID, clearBeforeDate, clearAfterDate);
+		}
+
+		sprintf(szTmp, "%.0f", value);		
+		long long MeterValue;
+		std::stringstream s_str2;
+		s_str2 << boost::to_string(value);
+		s_str2 >> MeterValue;
+
+		sprintf(szTmp, "%.0f", usage);		
+		long long MeterUsage;
+		std::stringstream s_str3;
+		s_str3 << boost::to_string(szTmp);
+		s_str3 >> MeterUsage;
+
+		//insert or replace record
+		result = safe_query(
+			"SELECT DeviceRowID FROM Meter "
+			"WHERE ((DeviceRowID=='%" PRIu64 "') AND (Date=='%q'))",
+			DeviceRowID, date
+		);
+		if (result.empty())
+		{
+			safe_query(
+				"INSERT INTO Meter (DeviceRowID, Value, Usage, Date) "
+				"VALUES ('%" PRIu64 "','%lld','%lld','%q')",
+				DeviceRowID, MeterValue, MeterUsage, date
+			);
+		}
+		else
+		{
+			safe_query(
+				"UPDATE Meter SET DeviceRowID='%" PRIu64 "', Value='%lld', Usage='%lld', Date='%q' "
```

Comment:
AFAICT we don't seem to do any validation of `date` at any point in the proceedings. The user could insert complete garbage into the database, couldn't they?

Reply:
That's true, I'm working on a different approach where the update will be possible on a specific managed device